File: src/AFK4.Customer.Web/src/screens/dashboard/liveSession.ts
export function elapsedSeconds(startedAtUtc: string, now: Date = new Date()): number {
  const started = Date.parse(startedAtUtc);
  if (Number.isNaN(started)) return 0;
  return Math.max(0, Math.floor((now.getTime() - started) / 1000));
}

export function projectRemainingSeconds(
  remainingAtFetch: number,
  fetchedAt: Date,
  now: Date = new Date()
): number {
  const drift = Math.floor((now.getTime() - fetchedAt.getTime()) / 1000);
  return Math.max(0, remainingAtFetch - drift);
}

export function formatClock(totalSeconds: number): string {
  const s = Math.max(0, Math.floor(totalSeconds));
  const hh = String(Math.floor(s / 3600)).padStart(2, '0');
  const mm = String(Math.floor((s % 3600) / 60)).padStart(2, '0');
  const ss = String(s % 60).padStart(2, '0');
  return `${hh}:${mm}:${ss}`;
}

File: src/afk4_customer_app/lib/dashboard/live_session.dart
/// Часы живой сессии. Отдельно от виджета, потому что это чистая арифметика, и ошибку в ней
/// дешевле поймать тестом, чем разглядыванием бегущих цифр.
library;

int elapsedSeconds(DateTime startedAtUtc, {required DateTime now}) {
  final seconds = now.difference(startedAtUtc).inSeconds;
  return seconds < 0 ? 0 : seconds;
}

/// Остаток по оплаченной сессии. Сервер отдаёт его на момент ответа, поэтому между опросами
/// остаток уменьшается сам — иначе цифра стояла бы на месте по полминуты и оживала рывком.
int projectRemainingSeconds(int remainingAtFetch, DateTime fetchedAt, {required DateTime now}) {
  final projected = remainingAtFetch - now.difference(fetchedAt).inSeconds;
  return projected < 0 ? 0 : projected;
}

String formatClock(int totalSeconds) {
  final s = totalSeconds < 0 ? 0 : totalSeconds;
  final hh = (s ~/ 3600).toString().padLeft(2, '0');
  final mm = ((s % 3600) ~/ 60).toString().padLeft(2, '0');
  final ss = (s % 60).toString().padLeft(2, '0');
  return '$hh:$mm:$ss';
}
